cpp/lib/answer/json.cpp

```diff
@@ -0,0 +1,313 @@
+        case JSONType::LONG:
+        case JSONType::DOUBLE: {
+            break;
+        }
+    }
+    memset(this, 0, sizeof(JSON));
+    _type = JSONType::NONE;
+}
+
+JSONType JSON::type() const {
+    return _type;
+}
+
+bool JSON::isMap() const {
+    return _type == JSONType::MAP;
+}
+
+bool JSON::isArray() const {
+    return _type == JSONType::ARRAY;
+}
+
+bool JSON::isBoolean() const {
+    return _type == JSONType::BOOLEAN;
+}
+
+bool JSON::isLong() const {
+    return _type == JSONType::LONG;
+}
+
+bool JSON::isDouble() const {
+    return _type == JSONType::DOUBLE;
+}
+
+bool JSON::isString() const {
+    return _type == JSONType::STRING;
+}
+
+const JSONMap& JSON::map() const {
+    if (!isMap()) throw Utils::exception(DriverError::INVALID_JSON_CAST, NAME(_type), NAME(JSONType::MAP));
+    return _map;
+}
+const JSONArray& JSON::array() const {
+    if (!isArray()) throw Utils::exception(DriverError::INVALID_JSON_CAST, NAME(_type), NAME(JSONType::ARRAY));
+    return _array;
+}
+const JSONBoolean& JSON::boolValue() const {
+    if (!isBoolean()) throw Utils::exception(DriverError::INVALID_JSON_CAST, NAME(_type), NAME(JSONType::BOOLEAN));
+    return _boolValue;
+}
+const JSONLong& JSON::longValue() const {
+    if (!isLong()) throw Utils::exception(DriverError::INVALID_JSON_CAST, NAME(_type), NAME(JSONType::LONG));
+    return _longValue;
+}
+const JSONDouble& JSON::doubleValue() const {
+    if (!isDouble()) throw Utils::exception(DriverError::INVALID_JSON_CAST, NAME(_type), NAME(JSONType::DOUBLE));
+    return _doubleValue;
+}
+const JSONString& JSON::stringValue() const {
+    if (!isString()) throw Utils::exception(DriverError::INVALID_JSON_CAST, NAME(_type), NAME(JSONType::STRING));
+    return _stringValue;
+}
+
+class JSONBuilder {
+   public:
+    JSONBuilder();
+    JSON build(const nlohmann::json& from);
+
+   private:
+    JSON buildMap(const nlohmann::json& from);
+    JSON buildArray(const nlohmann::json& from);
+};
+
+
+JSON JSON::parse(const std::string& string) {
+    return JSONBuilder().build(nlohmann::json::parse(string));
+}
+
+JSONBuilder::JSONBuilder() {}
+
+JSON JSONBuilder::build(const nlohmann::json& from) {
+    switch (from.type()) {
+        case nlohmann::json::value_t::boolean:
+            return JSON((JSONBoolean)from.get<bool>());
+        case nlohmann::json::value_t::number_integer:
+            return JSON((JSONLong)from.get<int>());
+        case nlohmann::json::value_t::number_unsigned:
+            return JSON((JSONLong)from.get<unsigned>());
+        case nlohmann::json::value_t::number_float:
+            return JSON((JSONDouble)from.get<unsigned>());
+        case nlohmann::json::value_t::string:
+            return JSON(from.get<std::string>());
+
+        case nlohmann::json::value_t::object:
+            return buildMap(from);
+        case nlohmann::json::value_t::array:
+            return buildArray(from);
+
+        default:
+            THROW_ILLEGAL_STATE;
+    }
+}
```

Comment:
I hit a wall when, because the recursive version of parse takes a parsed nlohmann::json object. I don't want to expose that type, so I'll still have an opaque forward declaration `ParsedJSON`, but can avoid the friendship.

I think this is a bit cleaner?